RecoTracker/MkFit/plugins/MkFitOutputConverter.cc

```diff
@@ -202,91 +194,78 @@ std::vector<const DetLayer*> MkFitOutputConverter::createDetLayers(const mkfit::
-    // TID+
-    off += 1;
-    set(off + 2, ttopo.tidDetId(side, 1, 0, 0, 0, 0));
-    set(off + 3, ttopo.tidDetId(side, 1, 0, 0, 0, 1));
-    set(off + 4, ttopo.tidDetId(side, 2, 0, 0, 0, 0));
-    set(off + 5, ttopo.tidDetId(side, 2, 0, 0, 0, 1));
-    set(off + 6, ttopo.tidDetId(side, 3, 0, 0, 0, 0));
-    set(off + 7, ttopo.tidDetId(side, 3, 0, 0, 0, 1));
-    // TEC
-    set(off + 8, ttopo.tecDetId(side, 1, 0, 0, 0, 0, 0));
-    set(off + 9, ttopo.tecDetId(side, 1, 0, 0, 0, 0, 1));
-    set(off + 10, ttopo.tecDetId(side, 2, 0, 0, 0, 0, 0));
-    set(off + 11, ttopo.tecDetId(side, 2, 0, 0, 0, 0, 1));
-    set(off + 12, ttopo.tecDetId(side, 3, 0, 0, 0, 0, 0));
-    set(off + 13, ttopo.tecDetId(side, 3, 0, 0, 0, 0, 1));
-    set(off + 14, ttopo.tecDetId(side, 4, 0, 0, 0, 0, 0));
-    set(off + 15, ttopo.tecDetId(side, 4, 0, 0, 0, 0, 1));
-    set(off + 16, ttopo.tecDetId(side, 5, 0, 0, 0, 0, 0));
-    set(off + 17, ttopo.tecDetId(side, 5, 0, 0, 0, 0, 1));
-    set(off + 18, ttopo.tecDetId(side, 6, 0, 0, 0, 0, 0));
-    set(off + 19, ttopo.tecDetId(side, 6, 0, 0, 0, 0, 1));
-    set(off + 20, ttopo.tecDetId(side, 7, 0, 0, 0, 0, 0));
-    set(off + 21, ttopo.tecDetId(side, 7, 0, 0, 0, 0, 1));
-    set(off + 22, ttopo.tecDetId(side, 8, 0, 0, 0, 0, 0));
-    set(off + 23, ttopo.tecDetId(side, 8, 0, 0, 0, 0, 1));
-    set(off + 24, ttopo.tecDetId(side, 9, 0, 0, 0, 0, 0));
-    set(off + 25, ttopo.tecDetId(side, 9, 0, 0, 0, 0, 1));
+  auto setBarrel = [&](int det, auto detIdFunc, int layer, int stereo) {
+    setDet(lnc.convertLayerNumber(det, layer, false, stereo, true), detIdFunc(layer, stereo));
   };
+  auto setForward = [&](int det, auto detIdFunc, int disk, int stereo) {
+    // minus side
+    setDet(lnc.convertLayerNumber(det, disk, false, stereo, false),
+           detIdFunc(static_cast<unsigned>(TrackerDetSide::NegEndcap), disk, stereo));
+    // plus side
+    setDet(lnc.convertLayerNumber(det, disk, false, stereo, true),
+           detIdFunc(static_cast<unsigned>(TrackerDetSide::PosEndcap), disk, stereo));
+  };
+
+  auto pxbDetId = [&ttopo](int layer, int stereo) { return ttopo.pxbDetId(layer, 0, 0); };
+  auto tibDetId = [&ttopo](int layer, int stereo) { return ttopo.tibDetId(layer, 0, 0, 0, 0, stereo); };
+  auto tobDetId = [&ttopo](int layer, int stereo) { return ttopo.tobDetId(layer, 0, 0, 0, stereo); };
 
-  // plus
-  off = 17 + 1;
-  setForward(2);
+  auto pxfDetId = [&ttopo](int side, int disk, int stereo) { return ttopo.pxfDetId(side, disk, 0, 0, 0); };
+  auto tidDetId = [&ttopo](int side, int disk, int stereo) { return ttopo.tidDetId(side, disk, 0, 0, 0, stereo); };
+  auto tecDetId = [&ttopo](int side, int disk, int stereo) { return ttopo.tecDetId(side, disk, 0, 0, 0, 0, stereo); };
 
-  // minus
-  off = 17 + 1 + 25 + 2;
-  setForward(1);
+  // TODO: detector structure currently hardcoded...
+  // BPix
+  for (int layer = 1; layer <= 4; ++layer) {
+    setBarrel(PixelSubdetector::PixelBarrel, pxbDetId, layer, 0);
```

Comment:
I'm still a bit puzzled why this needs to be so explicit and verbose.
can we just do something like
```c++
for (auto lay : tracker.allLayers()){
dets[lnc.convertLayerNumber(/*args*/)] = lay;
}
```
perhaps with a check that the layer is in the tracker

Reply:
Is there a better way to obtain the layer number from a `DetLayer` than via contained `GeomDet` (i.e. along `ttopo.layer(lay->basicComponents().front().geographicalId())`)?

Based on the implementation of `GeometricSearchTracker::idToLayer()` the `pixelBarrelLayers()`, `negPixelForwardLayers()` etc. are indexed by the layer number (`-1`) and thus going through `DetId` and `TrackerTopology` could in principle be avoided, but calling the detector-part-specific functions would naively lead to 9-fold duplication of the loop.

Or can it be assumed that the `DetLayer`s within a subdet and side are together and in order in `allLayers()`?

Reply:
I wasn't going to suggest to bypass the ```TrackerTopology```.
My comment was about avoiding explicit BPIX, FPIX, TIB, TOB, TID, TEC code blocs.
It seems possible to get DetId from a ```DetLayer``` (geographicalId). So, simply looping over ```allLayers()``` should be enough to fill the mkFit layer map.

Reply:
How about now?

Reply:
this is impressively better, at least an order of magnitude less code